File: packages/auth/src/frame.ts
import { NATIVE_AUTHORITY_PROTOCOL_VERSION } from "@recurs/contracts";

export const NATIVE_FRAME_MAGIC = 0x52_43_55_52;
export const NATIVE_FRAME_HEADER_BYTES = 16;
export const NATIVE_FRAME_MAX_PAYLOAD_BYTES = 8 * 1024 * 1024;
const NATIVE_FRAME_MAX_BYTES =
  NATIVE_FRAME_HEADER_BYTES + NATIVE_FRAME_MAX_PAYLOAD_BYTES;
const NativeUint8Array = Uint8Array;
const nativeTypedArrayPrototype = Object.getPrototypeOf(
  NativeUint8Array.prototype,
) as object;
const nativeBufferGetter = Object.getOwnPropertyDescriptor(
  nativeTypedArrayPrototype,
  "buffer",
)?.get as (this: Uint8Array) => ArrayBufferLike;
const nativeByteLengthGetter = Object.getOwnPropertyDescriptor(
  nativeTypedArrayPrototype,
  "byteLength",
)?.get as (this: Uint8Array) => number;
const nativeByteOffsetGetter = Object.getOwnPropertyDescriptor(
  nativeTypedArrayPrototype,
  "byteOffset",
)?.get as (this: Uint8Array) => number;
const nativeSet = NativeUint8Array.prototype.set;
const nativeFill = NativeUint8Array.prototype.fill;

export enum NativeMessageType {
  hello = 1,
  helloResult = 2,
  health = 3,
  healthResult = 4,
  cancel = 5,
  openAIOnboardingRequest = 6,
  openAIOnboardingBegun = 7,
  openAIOnboardingCatalogPage = 8,
  openAIOnboardingCommitted = 9,
  openAIOnboardingAborted = 10,
  openAIOnboardingReconciliation = 11,
  openAIOnboardingFailure = 12,
  openAIGenerationRequest = 13,
  openAIGenerationEvent = 14,
  openAIGenerationFailure = 15,
  safeFailure = 255,
}
Object.freeze(NativeMessageType);

export type NativeCodecErrorCode =
  | "invalid_frame"
  | "truncated_frame"
  | "decoder_finished"
  | "decoder_failed"
  | "invalid_field_table"
  | "invalid_field"
  | "invalid_message";

const nativeCodecErrorMessages: Readonly<Record<NativeCodecErrorCode, string>> =
  Object.freeze({
    invalid_frame: "Invalid native authority frame.",
    truncated_frame: "Truncated native authority frame.",
    decoder_finished: "Native authority decoder is finished.",
    decoder_failed: "Native authority decoder has failed.",
    invalid_field_table: "Invalid native authority field table.",
    invalid_field: "Invalid native authority field.",
    invalid_message: "Invalid native authority message.",
  });

export class NativeCodecError extends Error {
  readonly code: NativeCodecErrorCode;

  constructor(code: NativeCodecErrorCode) {
    super(nativeCodecErrorMessages[code]);
    this.name = "NativeCodecError";
    this.code = code;
  }
}

export function failNativeCodec(code: NativeCodecErrorCode): never {
  throw new NativeCodecError(code);
}

export function isNativeByteArray(value: unknown): value is Uint8Array {
  return value instanceof NativeUint8Array;
}

export function nativeByteStorage(value: Uint8Array): ArrayBufferLike {
  return Reflect.apply(nativeBufferGetter, value, []) as ArrayBufferLike;
}

export function nativeByteLength(value: Uint8Array): number {
  return Reflect.apply(nativeByteLengthGetter, value, []) as number;
}

export function nativeByteOffset(value: Uint8Array): number {
  return Reflect.apply(nativeByteOffsetGetter, value, []) as number;
}

export function nativeByteView(
  value: Uint8Array,
  begin = 0,
  end = nativeByteLength(value),
): Uint8Array {
  const length = nativeByteLength(value);
  if (
    !Number.isInteger(begin) ||
    !Number.isInteger(end) ||
    begin < 0 ||
    end < begin ||
    end > length
  ) {
    throw new RangeError("Invalid native byte view.");
  }
  return new NativeUint8Array(
    nativeByteStorage(value),
    nativeByteOffset(value) + begin,
    end - begin,
  );
}

export function copyNativeBytes(
  value: Uint8Array,
  begin = 0,
  end = nativeByteLength(value),
): Uint8Array {
  return new NativeUint8Array(nativeByteView(value, begin, end));
}

export function allocateNativeBytes(length: number): Uint8Array {
  return new NativeUint8Array(length);
}

export function writeNativeBytes(
  target: Uint8Array,
  source: Uint8Array,
  offset = 0,
): void {
  Reflect.apply(nativeSet, target, [source, offset]);
}

function zeroNativeBytes(
  target: Uint8Array,
  begin: number,
  end: number,
): void {
  Reflect.apply(nativeFill, target, [0, begin, end]);
}

export interface NativeFrame {
  readonly type: NativeMessageType;
  readonly requestId: number;
  readonly payload: Uint8Array;
}

function isNativeMessageType(value: number): value is NativeMessageType {
  switch (value) {
    case NativeMessageType.hello:
    case NativeMessageType.helloResult:
    case NativeMessageType.health:
    case NativeMessageType.healthResult:
    case NativeMessageType.cancel:
    case NativeMessageType.openAIOnboardingRequest:
    case NativeMessageType.openAIOnboardingBegun:
    case NativeMessageType.openAIOnboardingCatalogPage:
    case NativeMessageType.openAIOnboardingCommitted:
    case NativeMessageType.openAIOnboardingAborted:
    case NativeMessageType.openAIOnboardingReconciliation:
    case NativeMessageType.openAIOnboardingFailure:
    case NativeMessageType.openAIGenerationRequest:
    case NativeMessageType.openAIGenerationEvent:
    case NativeMessageType.openAIGenerationFailure:
    case NativeMessageType.safeFailure:
      return true;
    default:
      return false;
  }
}

function isU32(value: number): boolean {
  return Number.isInteger(value) && value >= 0 && value <= 0xffff_ffff;
}

function hasSharedByteStorage(value: Uint8Array): boolean {
  const storage = nativeByteStorage(value);
  return (
    typeof SharedArrayBuffer !== "undefined" &&
    storage instanceof SharedArrayBuffer
  );
}

function validateFrameParts(
  type: number,
  requestId: number,
  payloadLength: number,
): asserts type is NativeMessageType {
  if (
    !isNativeMessageType(type) ||
    !isU32(requestId) ||
    requestId === 0 ||
    !Number.isInteger(payloadLength) ||
    payloadLength < 0 ||
    payloadLength > NATIVE_FRAME_MAX_PAYLOAD_BYTES
  ) {
    failNativeCodec("invalid_frame");
  }
}

function createNativeFrame(
  type: NativeMessageType,
  requestId: number,
  payload: Uint8Array,
): NativeFrame {
  const storedPayload = copyNativeBytes(payload);
  return Object.freeze({
    type,
    requestId,
    get payload(): Uint8Array {
      return copyNativeBytes(storedPayload);
    },
  });
}

export function encodeNativeFrame(frame: NativeFrame): Uint8Array {
  try {
    const type = frame.type;
    const requestId = frame.requestId;
    const inputPayload = frame.payload;
    if (
      !isNativeByteArray(inputPayload) ||
      hasSharedByteStorage(inputPayload)
    ) {
      failNativeCodec("invalid_frame");
    }
    const payload = copyNativeBytes(inputPayload);
    const payloadLength = nativeByteLength(payload);
    validateFrameParts(type, requestId, payloadLength);

    const encoded = allocateNativeBytes(
      NATIVE_FRAME_HEADER_BYTES + payloadLength,
    );
    const view = new DataView(nativeByteStorage(encoded));
    view.setUint32(0, NATIVE_FRAME_MAGIC, false);
    view.setUint16(4, NATIVE_AUTHORITY_PROTOCOL_VERSION, false);
    view.setUint16(6, type, false);
    view.setUint32(8, payloadLength, false);
    view.setUint32(12, requestId, false);
    writeNativeBytes(encoded, payload, NATIVE_FRAME_HEADER_BYTES);
    return encoded;
  } catch {
    failNativeCodec("invalid_frame");
  }
}

interface ParsedHeader {
  readonly type: NativeMessageType;
  readonly requestId: number;
  readonly payloadLength: number;
  readonly frameLength: number;
}

function parseHeader(bytes: Uint8Array, offset: number): ParsedHeader {
  const view = new DataView(
    nativeByteStorage(bytes),
    nativeByteOffset(bytes) + offset,
    NATIVE_FRAME_HEADER_BYTES,
  );
  const magic = view.getUint32(0, false);
  const protocolVersion = view.getUint16(4, false);
  const type = view.getUint16(6, false);
  const payloadLength = view.getUint32(8, false);
  const requestId = view.getUint32(12, false);

  if (
    magic !== NATIVE_FRAME_MAGIC ||
    protocolVersion !== NATIVE_AUTHORITY_PROTOCOL_VERSION
  ) {
    failNativeCodec("invalid_frame");
  }
  validateFrameParts(type, requestId, payloadLength);
  return Object.freeze({
    type,
    requestId,
    payloadLength,
    frameLength: NATIVE_FRAME_HEADER_BYTES + payloadLength,
  });
}

type DecoderState = "open" | "finished" | "failed";

export class NativeFrameDecoder {
  #state: DecoderState = "open";
  #buffer: Uint8Array | undefined;
  #bufferedLength = 0;

  push(chunk: Uint8Array): readonly NativeFrame[] {
    this.#ensureOpen();
    const frames: NativeFrame[] = [];
    try {
      if (
        !isNativeByteArray(chunk) ||
        hasSharedByteStorage(chunk)
      ) {
        failNativeCodec("invalid_frame");
      }
      const inputLength = nativeByteLength(chunk);
      let offset = 0;
      while (offset < inputLength || this.#bufferedLength > 0) {
        if (this.#bufferedLength > 0) {
          if (this.#bufferedLength < NATIVE_FRAME_HEADER_BYTES) {
            const take = Math.min(
              NATIVE_FRAME_HEADER_BYTES - this.#bufferedLength,
              inputLength - offset,
            );
            if (take > 0) {
              this.#appendBuffered(chunk, offset, take);
              offset += take;
            }
            if (this.#bufferedLength < NATIVE_FRAME_HEADER_BYTES) {
              break;
            }
          }

          const buffer = this.#buffer;
          if (buffer === undefined) {
            failNativeCodec("invalid_frame");
          }
          const header = parseHeader(buffer, 0);
          const take = Math.min(
            header.frameLength - this.#bufferedLength,
            inputLength - offset,
          );
          if (take > 0) {
            this.#appendBuffered(chunk, offset, take);
            offset += take;
          }
          if (this.#bufferedLength < header.frameLength) {
            break;
          }
          frames.push(createNativeFrame(
            header.type,
            header.requestId,
            nativeByteView(
              buffer,
              NATIVE_FRAME_HEADER_BYTES,
              header.frameLength,
            ),
          ));
          this.#clearBuffered(false);
          continue;
        }

        const remaining = inputLength - offset;
        if (remaining < NATIVE_FRAME_HEADER_BYTES) {
          this.#appendBuffered(chunk, offset, remaining);
          offset = inputLength;
          break;
        }

        const header = parseHeader(chunk, offset);
        if (remaining < header.frameLength) {
          this.#appendBuffered(chunk, offset, remaining);
          offset = inputLength;
          break;
        }
        frames.push(createNativeFrame(
          header.type,
          header.requestId,
          nativeByteView(
            chunk,
            offset + NATIVE_FRAME_HEADER_BYTES,
            offset + header.frameLength,
          ),
        ));
        offset += header.frameLength;
      }
    } catch {
      this.#poison();
      failNativeCodec("invalid_frame");
    }

    return Object.freeze(frames);
  }

  finish(): void {
    this.#ensureOpen();
    if (this.#bufferedLength !== 0) {
      this.#poison();
      failNativeCodec("truncated_frame");
    }
    this.#clearBuffered(true);
    this.#state = "finished";
  }

  #appendBuffered(source: Uint8Array, offset: number, length: number): void {
    if (
      !Number.isInteger(offset) ||
      !Number.isInteger(length) ||
      offset < 0 ||
      length < 0 ||
      offset + length > nativeByteLength(source) ||
      this.#bufferedLength + length > NATIVE_FRAME_MAX_BYTES
    ) {
      failNativeCodec("invalid_frame");
    }
    if (length === 0) {
      return;
    }
    const buffer = this.#buffer ?? allocateNativeBytes(NATIVE_FRAME_MAX_BYTES);
    this.#buffer = buffer;
    writeNativeBytes(
      buffer,
      nativeByteView(source, offset, offset + length),
      this.#bufferedLength,
    );
    this.#bufferedLength += length;
  }

  #clearBuffered(release: boolean): void {
    if (this.#buffer !== undefined && this.#bufferedLength > 0) {
      zeroNativeBytes(this.#buffer, 0, this.#bufferedLength);
    }
    this.#bufferedLength = 0;
    if (release) {
      this.#buffer = undefined;
    }
  }

  #ensureOpen(): void {
    if (this.#state === "finished") {
      failNativeCodec("decoder_finished");
    }
    if (this.#state === "failed") {
      failNativeCodec("decoder_failed");
    }
  }

  #poison(): void {
    this.#clearBuffered(true);
    this.#state = "failed";
  }
}

File: native/macos/Sources/RecursNativeProtocol/Frame.swift
import Foundation

public let nativeAuthorityProtocolVersion: UInt16 = 1
public let nativeFrameMagic: UInt32 = 0x52_43_55_52
public let nativeFrameHeaderByteCount = 16
public let nativeFrameMaximumPayloadByteCount = 8 * 1024 * 1024

private let nativeFrameMaximumByteCount =
  nativeFrameHeaderByteCount + nativeFrameMaximumPayloadByteCount

public enum NativeMessageType: UInt16, CaseIterable, Sendable {
  case hello = 1
  case helloResult = 2
  case health = 3
  case healthResult = 4
  case cancel = 5
  case openAIOnboardingRequest = 6
  case openAIOnboardingBegun = 7
  case openAIOnboardingCatalogPage = 8
  case openAIOnboardingCommitted = 9
  case openAIOnboardingAborted = 10
  case openAIOnboardingReconciliation = 11
  case openAIOnboardingFailure = 12
  case openAIGenerationRequest = 13
  case openAIGenerationEvent = 14
  case openAIGenerationFailure = 15
  case safeFailure = 255
}

public enum NativeProtocolError: Error, Equatable, Sendable, LocalizedError {
  case invalidFrame
  case truncatedFrame
  case decoderFinished
  case decoderFailed
  case invalidFieldTable
  case invalidField
  case invalidMessage

  public var errorDescription: String? {
    switch self {
    case .invalidFrame:
      "Invalid native authority frame."
    case .truncatedFrame:
      "Truncated native authority frame."
    case .decoderFinished:
      "Native authority decoder is finished."
    case .decoderFailed:
      "Native authority decoder has failed."
    case .invalidFieldTable:
      "Invalid native authority field table."
    case .invalidField:
      "Invalid native authority field."
    case .invalidMessage:
      "Invalid native authority message."
    }
  }
}

public struct NativeFrame: Equatable, Sendable {
  public let type: NativeMessageType
  public let requestID: UInt32

  private let payloadStorage: [UInt8]

  public var payload: Data {
    Data(payloadStorage)
  }

  public init(
    type: NativeMessageType,
    requestID: UInt32,
    payload: Data
  ) throws {
    guard requestID != 0, payload.count <= nativeFrameMaximumPayloadByteCount else {
      throw NativeProtocolError.invalidFrame
    }
    self.type = type
    self.requestID = requestID
    self.payloadStorage = Array(payload)
  }

  internal init(
    validatedType type: NativeMessageType,
    requestID: UInt32,
    payloadBytes: [UInt8]
  ) {
    self.type = type
    self.requestID = requestID
    self.payloadStorage = Array(payloadBytes)
  }

  public func encoded() throws -> Data {
    guard requestID != 0, payloadStorage.count <= nativeFrameMaximumPayloadByteCount else {
      throw NativeProtocolError.invalidFrame
    }

    var bytes: [UInt8] = []
    bytes.reserveCapacity(nativeFrameHeaderByteCount + payloadStorage.count)
    nativeAppendUInt32(nativeFrameMagic, to: &bytes)
    nativeAppendUInt16(nativeAuthorityProtocolVersion, to: &bytes)
    nativeAppendUInt16(type.rawValue, to: &bytes)
    nativeAppendUInt32(UInt32(payloadStorage.count), to: &bytes)
    nativeAppendUInt32(requestID, to: &bytes)
    bytes.append(contentsOf: payloadStorage)
    return Data(bytes)
  }
}

public struct NativeFrameDecoder: Sendable {
  private enum State: Sendable {
    case open
    case finished
    case failed
  }

  private var state = State.open
  private var bufferedBytes: [UInt8] = []
  private var bufferedHeader: NativeFrameHeader?

  public init() {}

  package var isAwaitingFrameCompletion: Bool {
    guard case .open = state else {
      return false
    }
    return !bufferedBytes.isEmpty
  }

  public mutating func push(_ chunk: Data) throws -> [NativeFrame] {
    try ensureOpen()
    guard !chunk.isEmpty else {
      return []
    }

    do {
      let input = Array(chunk)
      var frames: [NativeFrame] = []
      var offset = 0

      while offset < input.count || !bufferedBytes.isEmpty {
        if !bufferedBytes.isEmpty {
          if bufferedBytes.count < nativeFrameHeaderByteCount {
            let needed = nativeFrameHeaderByteCount - bufferedBytes.count
            let available = input.count - offset
            let take = min(needed, available)
            try appendBuffered(input, offset: offset, count: take)
            offset += take
            if bufferedBytes.count < nativeFrameHeaderByteCount {
              break
            }
          }

          let header: NativeFrameHeader
          if let existingHeader = bufferedHeader {
            header = existingHeader
          } else {
            header = try nativeParseFrameHeader(bufferedBytes, offset: 0)
            bufferedHeader = header
          }

          let needed = header.frameByteCount - bufferedBytes.count
          let available = input.count - offset
          let take = min(needed, available)
          try appendBuffered(input, offset: offset, count: take)
          offset += take
          if bufferedBytes.count < header.frameByteCount {
            break
          }

          frames.append(
            try nativeMakeFrame(from: bufferedBytes, offset: 0, header: header)
          )
          clearBuffered(releaseStorage: false)
          continue
        }

        let remaining = input.count - offset
        if remaining < nativeFrameHeaderByteCount {
          try appendBuffered(input, offset: offset, count: remaining)
          offset = input.count
          break
        }

        let header = try nativeParseFrameHeader(input, offset: offset)
        if remaining < header.frameByteCount {
          try appendBuffered(input, offset: offset, count: remaining)
          bufferedHeader = header
          offset = input.count
          break
        }

        frames.append(try nativeMakeFrame(from: input, offset: offset, header: header))
        offset += header.frameByteCount
      }

      return frames
    } catch {
      poison()
      throw NativeProtocolError.invalidFrame
    }
  }

  public mutating func finish() throws {
    try ensureOpen()
    guard bufferedBytes.isEmpty else {
      poison()
      throw NativeProtocolError.truncatedFrame
    }
    clearBuffered(releaseStorage: true)
    state = .finished
  }

  private func ensureOpen() throws {
    switch state {
    case .open:
      return
    case .finished:
      throw NativeProtocolError.decoderFinished
    case .failed:
      throw NativeProtocolError.decoderFailed
    }
  }

  private mutating func appendBuffered(
    _ source: [UInt8],
    offset: Int,
    count: Int
  ) throws {
    guard
      offset >= 0,
      count >= 0,
      offset <= source.count,
      count <= source.count - offset,
      bufferedBytes.count <= nativeFrameMaximumByteCount,
      count <= nativeFrameMaximumByteCount - bufferedBytes.count
    else {
      throw NativeProtocolError.invalidFrame
    }
    guard count > 0 else {
      return
    }

    if bufferedBytes.isEmpty {
      bufferedBytes.reserveCapacity(nativeFrameMaximumByteCount)
    }
    bufferedBytes.append(contentsOf: source[offset..<(offset + count)])
  }

  private mutating func clearBuffered(releaseStorage: Bool) {
    for index in bufferedBytes.indices {
      bufferedBytes[index] = 0
    }
    bufferedBytes.removeAll(keepingCapacity: !releaseStorage)
    bufferedHeader = nil
  }

  private mutating func poison() {
    clearBuffered(releaseStorage: true)
    state = .failed
  }
}

private struct NativeFrameHeader: Sendable {
  let type: NativeMessageType
  let requestID: UInt32
  let payloadByteCount: Int
  let frameByteCount: Int
}

private func nativeParseFrameHeader(
  _ bytes: [UInt8],
  offset: Int
) throws -> NativeFrameHeader {
  guard
    offset >= 0,
    offset <= bytes.count,
    nativeFrameHeaderByteCount <= bytes.count - offset,
    let magic = nativeReadUInt32(bytes, at: offset),
    let protocolVersion = nativeReadUInt16(bytes, at: offset + 4),
    let rawType = nativeReadUInt16(bytes, at: offset + 6),
    let rawPayloadByteCount = nativeReadUInt32(bytes, at: offset + 8),
    let requestID = nativeReadUInt32(bytes, at: offset + 12),
    magic == nativeFrameMagic,
    protocolVersion == nativeAuthorityProtocolVersion,
    let type = NativeMessageType(rawValue: rawType),
    requestID != 0,
    rawPayloadByteCount <= UInt32(nativeFrameMaximumPayloadByteCount)
  else {
    throw NativeProtocolError.invalidFrame
  }

  let payloadByteCount = Int(rawPayloadByteCount)
  return NativeFrameHeader(
    type: type,
    requestID: requestID,
    payloadByteCount: payloadByteCount,
    frameByteCount: nativeFrameHeaderByteCount + payloadByteCount
  )
}

private func nativeMakeFrame(
  from bytes: [UInt8],
  offset: Int,
  header: NativeFrameHeader
) throws -> NativeFrame {
  guard
    offset >= 0,
    offset <= bytes.count,
    header.frameByteCount <= bytes.count - offset
  else {
    throw NativeProtocolError.invalidFrame
  }

  let payloadOffset = offset + nativeFrameHeaderByteCount
  let payloadEnd = payloadOffset + header.payloadByteCount
  let payloadBytes = Array(bytes[payloadOffset..<payloadEnd])
  return NativeFrame(
    validatedType: header.type,
    requestID: header.requestID,
    payloadBytes: payloadBytes
  )
}

internal func nativeReadUInt16(_ bytes: [UInt8], at offset: Int) -> UInt16? {
  guard offset >= 0, offset <= bytes.count, 2 <= bytes.count - offset else {
    return nil
  }
  return (UInt16(bytes[offset]) << 8)
    | UInt16(bytes[offset + 1])
}

internal func nativeReadUInt32(_ bytes: [UInt8], at offset: Int) -> UInt32? {
  guard offset >= 0, offset <= bytes.count, 4 <= bytes.count - offset else {
    return nil
  }
  return (UInt32(bytes[offset]) << 24)
    | (UInt32(bytes[offset + 1]) << 16)
    | (UInt32(bytes[offset + 2]) << 8)
    | UInt32(bytes[offset + 3])
}

internal func nativeReadUInt64(_ bytes: [UInt8], at offset: Int) -> UInt64? {
  guard offset >= 0, offset <= bytes.count, 8 <= bytes.count - offset else {
    return nil
  }
  var value: UInt64 = 0
  for index in 0..<8 {
    value = (value << 8) | UInt64(bytes[offset + index])
  }
  return value
}

internal func nativeAppendUInt16(_ value: UInt16, to bytes: inout [UInt8]) {
  bytes.append(UInt8((value >> 8) & 0xff))
  bytes.append(UInt8(value & 0xff))
}

internal func nativeAppendUInt32(_ value: UInt32, to bytes: inout [UInt8]) {
  bytes.append(UInt8((value >> 24) & 0xff))
  bytes.append(UInt8((value >> 16) & 0xff))
  bytes.append(UInt8((value >> 8) & 0xff))
  bytes.append(UInt8(value & 0xff))
}

internal func nativeAppendUInt64(_ value: UInt64, to bytes: inout [UInt8]) {
  for shift in stride(from: 56, through: 0, by: -8) {
    bytes.append(UInt8((value >> UInt64(shift)) & 0xff))
  }
}
